wrapper/src/main/java/software/amazon/jdbc/dialect/DialectManager.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc.dialect;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.TimeUnit;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import software.amazon.jdbc.AwsWrapperProperty;
+import software.amazon.jdbc.HostSpec;
+import software.amazon.jdbc.util.CacheMap;
+import software.amazon.jdbc.util.ConnectionUrlParser;
+import software.amazon.jdbc.util.Messages;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+import software.amazon.jdbc.util.Utils;
+
+public class DialectManager implements DialectProvider {
+
+  protected static DialectProvider customDialectProvider;
+
+  public static final AwsWrapperProperty DIALECT = new AwsWrapperProperty(
+      "wrapperDialect", "",
+      "A unique identifier for the supported database dialect.");
+
+  /**
+   * Every Dialect implementation SHOULD BE stateless!!!
+   * Dialect objects are shared between different connections.
+   */
+  protected static final Map<String, Dialect> knownDialectsByCode =
+      new HashMap<String, Dialect>() {
+        {
+          put(DialectCodes.MYSQL, new MysqlDialect());
+          put(DialectCodes.PG, new PgDialect());
+          put(DialectCodes.MARIADB, new MariaDbDialect());
+          put(DialectCodes.RDS_MYSQL, new RdsMysqlDialect());
+          put(DialectCodes.RDS_PG, new RdsPgDialect());
+          put(DialectCodes.AURORA_MYSQL, new AuroraMysqlDialect());
+          put(DialectCodes.AURORA_PG, new AuroraPgDialect());
+          put(DialectCodes.UNKNOWN, new UnknownDialect());
+        }
+      };
+
+  protected static final long ENDPOINT_CACHE_EXPIRATION = TimeUnit.MINUTES.toNanos(30);
+
+  // Map of host name, or url, by dialect code.
+  protected static final CacheMap<String, String> knownEndpointDialects = new CacheMap<>();
+
+  private final RdsUtils rdsHelper = new RdsUtils();
+  private final ConnectionUrlParser connectionUrlParser = new ConnectionUrlParser();
+  private boolean canUpdate = false;
+  private Dialect dialect;
```

Comment:
```suggestion
  private Dialect dialect = null;
```
Then you don't have to do `this.dialect = null;` in getDialect()